# SMS Spam Filter 

    We will build a Spam Filter for classifying SMS messages as spam or non-spam using Naive Bayes algorithm. For training   algorithm we will use the file, The UCI Machine Learning Repository: https://archive.ics.uci.edu/ml/datasets/sms+spam+collection . The project is done with guidance of Dataquest community.
    There are 5000+ SMS messages already classified by human as spam or non-spam in this file. We will divide the set of all data into two parts at the ratio 80% — 20 %. The first bigger set will be used for training of algorithm, the second part we will use to check how the program works, for testing purpose only. 
    Let us look at the file:

In [112]:
import pandas as pd
import numpy as np
file = pd.read_csv("SMSSpamCollection", sep='\t',header=None, names=['Label', 'SMS'])

In [113]:
file.head(10)

Label                                                SMS
0   ham  Go until jurong point, crazy.. Available only ...
1   ham                      Ok lar... Joking wif u oni...
2  spam  Free entry in 2 a wkly comp to win FA Cup fina...
3   ham  U dun say so early hor... U c already then say...
4   ham  Nah I don't think he goes to usf, he lives aro...
5  spam  FreeMsg Hey there darling it's been 3 week's n...
6   ham  Even my brother is not like to speak with me. ...
7   ham  As per your request 'Melle Melle (Oru Minnamin...
8  spam  WINNER!! As a valued network customer you have...
9  spam  Had your mobile 11 months or more? U R entitle...

In [114]:
file.shape

(5572, 2)

In [115]:
len(file.index)

5572

In [116]:
percent_spam = len(file[file['Label']=='spam']) /len(file.index) *100
percent_ham = len(file[file['Label']=='ham']) /len(file.index) *100

In [117]:
print('Percentage of spam messages out of all = ', round(percent_spam,1), ' %')

Percentage of spam messages out of all =  13.4  %


In [118]:
print('Percentage of non-spam messages out of all = ', round(percent_ham,1), ' %')

Percentage of non-spam messages out of all =  86.6  %


Let us divide the data 80%-20%: one set for training, the other data will left for final testing only:

80% for training purpose:

In [119]:
training = file.sample(frac=0.8,axis=0,random_state=1).copy()

Command sample() helped us to choose data in random order. 
    Let us safe all initial indices from original file for rows of Training set in one list, so as all other indices use for creating the second file for testing.

In [120]:
ind = training.index.tolist()

Now we reset old indices for new ones in the training set:

In [121]:
training = training.reset_index(drop=True)
training.head()

Label                                                SMS
0   ham                       Yep, by the pretty sculpture
1   ham      Yes, princess. Are you going to make me moan?
2   ham                         Welp apparently he retired
3   ham                                            Havent.
4   ham  I forgot 2 ask ü all smth.. There's a card on ...

The number of all rows in initial file:

In [122]:
indices = file.index.tolist()

Let us create a new list index_testing with indices of all elements which not get into the training data set. 
We will use these indices to create the second dataset (20% of all for checking how accurately the created spam filter works only).

In [123]:
index_testing =[]
for row in range(len(indices)):
    if row not in ind:
        index_testing.append(row)
print(len(index_testing))

1114


Let us check do the lengths of training and checking datasets match with initial set of SMS messages:

In [124]:
print(len(file), len(training)+len(testing))

5572 5572


Yes, everything is correct in terms of size. Let us check that indices are save in random order:

In [125]:
print(index_testing[6:23])

[31, 42, 43, 46, 52, 61, 63, 69, 70, 71, 77, 84, 85, 95, 96, 97, 105]


Let us create a dictionary with chosen rows and convert it into data frame:

In [126]:
li = {}
for k in index_testing:
    l = file.loc[k]
    li[k]=l
li

{2: Label                                                 spam
 SMS      Free entry in 2 a wkly comp to win FA Cup fina...
 Name: 2, dtype: object,
 14: Label                                    ham
 SMS      I HAVE A DATE ON SUNDAY WITH WILL!!
 Name: 14, dtype: object,
 15: Label                                                 spam
 SMS      XXXMobileMovieClub: To use your credit, click ...
 Name: 15, dtype: object,
 18: Label                                                  ham
 SMS      Fine if that’s the way u feel. That’s the way ...
 Name: 18, dtype: object,
 20: Label                                          ham
 SMS      Is that seriously how you spell his name?
 Name: 20, dtype: object,
 29: Label                                                  ham
 SMS      Ahhh. Work. I vaguely remember that! What does...
 Name: 29, dtype: object,
 31: Label                                                  ham
 SMS      Yeah he got in at 2 and was v apologetic. n ha...
 Name: 31, dtype: obje

In [127]:
testing = pd.DataFrame(li).transpose().reset_index(drop=True)


In [128]:
testing

Label                                                SMS
0     spam  Free entry in 2 a wkly comp to win FA Cup fina...
1      ham                I HAVE A DATE ON SUNDAY WITH WILL!!
2     spam  XXXMobileMovieClub: To use your credit, click ...
3      ham  Fine if that’s the way u feel. That’s the way ...
4      ham          Is that seriously how you spell his name?
...    ...                                                ...
1109   ham         Ic. There are a lotta childporn cars then.
1110   ham  You know, wot people wear. T shirts, jumpers, ...
1111   ham  Have a safe trip to Nigeria. Wish you happines...
1112   ham  Yeh. Indians was nice. Tho it did kane me off ...
1113   ham                                Ard 6 like dat lor.

[1114 rows x 2 columns]

In [129]:
training

Label                                                SMS
0      ham                       Yep, by the pretty sculpture
1      ham      Yes, princess. Are you going to make me moan?
2      ham                         Welp apparently he retired
3      ham                                            Havent.
4      ham  I forgot 2 ask ü all smth.. There's a card on ...
...    ...                                                ...
4453   ham  Sorry, I'll call later in meeting any thing re...
4454   ham  Babe! I fucking love you too !! You know? Fuck...
4455  spam  U've been selected to stay in 1 of 250 top Bri...
4456   ham  Hello my boytoy ... Geeee I miss you already a...
4457   ham                           Wherre's my boytoy ? :-(

[4458 rows x 2 columns]

So, as the result we have two datasets which divided our initial data into 80-20% ratio. 
    Let us check ratios of spam to non-spam messages in each of datasets.

In [130]:
percent_spam_train = len(training[training['Label']=='spam']) /len(training.index) *100
percent_ham_train = len(training[training['Label']=='ham'])/len(training.index) *100
print('Percentage of spam messages in Training set = ', round(percent_spam_train,1), ' %')
print('Percentage of non-spam messages in Training set = ', round(percent_ham_train,1), ' %')

Percentage of spam messages in Training set =  13.5  %
Percentage of non-spam messages in Training set =  86.5  %


In [131]:
percent_spam_test = len(testing[testing['Label']=='spam']) /len(testing.index) *100
percent_ham_test = len(testing[testing['Label']=='ham'])/len(testing.index) *100
print('Percentage of spam messages in Testing set = ', round(percent_spam_test,1), ' %')
print('Percentage of non-spam messages in Testing set = ', round(percent_ham_test,1), ' %')

Percentage of spam messages in Testing set =  13.2  %
Percentage of non-spam messages in Testing set =  86.8  %


In [132]:
testing['Label'].value_counts(normalize=True)

ham     0.868043
spam    0.131957
Name: Label, dtype: float64

The ratios of spam to non-spam messages are almost the same for all three datasets (initial, training set and testing). 
So, we satisfy how we splitted up dataset in terms of quantity. So, we can start clean the data in both datasets.

# Cleaning the data 

In [133]:
training['SMS'] = training['SMS'].replace('\W', ' ')
training['SMS'] = training['SMS'].str.lower()

In [134]:
testing[4:8]

Label                                                SMS
4   ham          Is that seriously how you spell his name?
5   ham  Ahhh. Work. I vaguely remember that! What does...
6   ham  Yeah he got in at 2 and was v apologetic. n ha...
7  spam  07732584351 - Rodger Burns - MSG = We tried to...

# Total Number of Words

Let us calculate total number of words used in training set. It is so called vocabulary of the set.

In [135]:
training['SMS'] = training['SMS'].str.split()
voc = []
for row in training['SMS']:
    for word in row:
        voc.append(word)
            
print(voc[:100])
vocabulary = list(set(voc))
print(vocabulary[:50])

['yep,', 'by', 'the', 'pretty', 'sculpture', 'yes,', 'princess.', 'are', 'you', 'going', 'to', 'make', 'me', 'moan?', 'welp', 'apparently', 'he', 'retired', 'havent.', 'i', 'forgot', '2', 'ask', 'ü', 'all', 'smth..', "there's", 'a', 'card', 'on', 'da', 'present', 'lei...', 'how?', 'ü', 'all', 'want', '2', 'write', 'smth', 'or', 'sign', 'on', 'it?', 'ok', 'i', 'thk', 'i', 'got', 'it.', 'then', 'u', 'wan', 'me', '2', 'come', 'now', 'or', 'wat?', 'i', 'want', 'kfc', 'its', 'tuesday.', 'only', 'buy', '2', 'meals', 'only', '2.', 'no', 'gravy.', 'only', '2', 'mark.', '2!', 'no', 'dear', 'i', 'was', 'sleeping', ':-p', 'ok', 'pa.', 'nothing', 'problem:-)', 'ill', 'be', 'there', 'on', '&lt;#&gt;', 'ok.', 'my', 'uncles', 'in', 'atlanta.', 'wish', 'you', 'guys', 'a']
['8.attractive', 'optout', '*sighs*', '6hl', 'smiling.', '"oh', 'free?', 'xmas!', 'way', 'sk38xh.', 'hello:"you', 'chat...', 'remember.', 'tiny', 'june..', '"mix"', 'flat,', 'care,whoever', 'actor.', 'tomeandsaid,this', '&lt;decimal&

Next step is to create a dictionary which shows how many time each unique word was used in the SMS:

In [136]:
word_counts_per_sms = {unique_word: [0] * len(training['SMS']) for unique_word in vocabulary}

for index, sms in enumerate(training['SMS']):
    for word in sms:
        word_counts_per_sms[word][index] += 1

Let convert the dictionary into a table. It will be used either for calculating all words used in single SMS either in all set. 

In [137]:
table_training = pd.DataFrame(data=word_counts_per_sms, columns=vocabulary)


In [138]:
table_training.iloc[100:110,250:275]

club4  rolled  successfully.  150p/tone.  msg:  apartment  08452810073  \
100      0       0              0           0     0          0            0   
101      0       0              0           0     0          0            0   
102      0       0              0           0     0          0            0   
103      0       0              0           0     0          0            0   
104      0       0              0           0     0          0            0   
105      0       0              0           0     0          0            0   
106      0       0              0           0     0          0            0   
107      0       0              0           0     0          0            0   
108      0       0              0           0     0          0            0   
109      0       0              0           0     0          0            0   

     panren  big!  umma.  ...  store  apo  luck!  burger  someone,  followed  \
100       0     0      0  ...      0    0      0       0         0         0   
101       0     0      0  ...      0    0      0       0         0         0   
102       0     0      0  ...      0    0      0       0         0         0   
103       0     0      0  ...      0    0      0       0         0         0   
104       0     0      0  ...      0    0      0       0         0         0   
105       0     0      0  ...      0    0      0       0         0         0   
106       0     0      0  ...      0    0      0       0         0         0   
107       0     0      0  ...      0    0      0       0         0         0   
108       0     0      0  ...      0    0      0       0         0         0   
109       0     0      0  ...      0    0      0       0         0         0   

     predict  a-?  normal  bbq  
100        0    0       0    0  
101        0    0       0    0  
102        0    0       0    0  
103        0    0       0    0  
104        0    0       0    0  
105        0    0       0    0  
106        0    0       0    0  
107        0    0       0    0  
108        0    0       0    0  
109        0    0       0    0  

[10 rows x 25 columns]

In [139]:
training_and_words = pd.concat([training, table_training],axis=1)

and merge the table with training set. 

In [140]:
training_and_words = pd.concat([training, table_training],axis=1)

Let us check, does the table correctly reflects the presence of words in the messages:

In [141]:
training_and_words.loc[99:103,['SMS','sorry','juicy','what','your','meet','good','london','my']]

SMS  sorry  juicy  what  \
99   [lul, im, gettin, some, juicy, gossip, at, the...      0      1     0   
100  [sorry, man,, my, stash, ran, dry, last, night...      1      0     0   
101  [well, good, morning, mr, ., hows, london, tre...      0      0     0   
102  [hey..., what, time, is, your, driving, on, fr...      0      0     1   
103                        [meet, after, lunch, la...]      0      0     0   

     your  meet  good  london  my  
99      0     0     0       0   0  
100     0     0     0       0   1  
101     0     0     1       1   0  
102     1     0     0       0   0  
103     0     1     0       0   0

Probabilities of getting the Spam ans Non-Spam messages are calculated below:

In [142]:
p_spam = len(training_and_words[training_and_words['Label']=='spam'])/len(training_and_words)
p_non_spam = 1 - p_spam

Let us separate the Spam messages only from training set into a separate table:

In [143]:
spam_base = training_and_words[training_and_words['Label']=='spam'].copy()

In [144]:
list_spam = spam_base['SMS'].tolist()

In [145]:
Num_spam_words = []
for row in list_spam:
    for word in row:
        Num_spam_words.append(word)
print('Total number of words in spam messages =',len(Num_spam_words))
n_spam = len(Num_spam_words)

Total number of words in spam messages = 14257


Let us repeat the procedure for non-spam messages:  we will calculate the number of unique words in non-spam messages.

In [146]:
non_spam_base = training_and_words[training_and_words['Label']=='ham'].copy()
list_non_spam_SMS = non_spam_base['SMS'].tolist()

In [147]:
Num_NON_spam_words = []
for row in list_non_spam_SMS:
    for word in row:
        Num_NON_spam_words.append(word)
print('Total number of words in NON spam messages =',len(Num_NON_spam_words))
n_non_spam = len(Num_NON_spam_words)

Total number of words in NON spam messages = 55376


These three parameters are needed for calculation of propabilities to meet each word in Spam  and non-Spam messages:
n_non_spam -  number of words in non-spam messages
n_spam - number of words in spam messages
N_voc - total number of unique words into set

In [148]:
N_voc = len(list(set(voc)))  # Length of the vocabulary - total number of unique words
print(n_spam, n_non_spam, N_voc)  

14257 55376 11860


In [149]:
alpha = 1 # smoothing parameter in Bayes Algorithm

Now we know all factors to calculate probabilities to encounter each word in spam or non-spam message. 
Let us put info on all words into dictionaries.


In [150]:
dict_word_in_spam = {each_word: 0 for each_word in vocabulary}

In [151]:
dict_word_in_non_spam = {each_word: 0 for each_word in vocabulary}

In [152]:
for each_word in vocabulary:
    num_s = spam_base[each_word].sum()
    dict_word_in_spam[each_word] = (num_s+alpha)/(n_spam + alpha * N_voc)
    num_h = non_spam_base[each_word].sum()
    dict_word_in_non_spam[each_word] = (num_h + alpha)/(n_non_spam + alpha*N_voc)

In [153]:
#Finaly, we have all to create the Spam Filter:

In [154]:
import re

def spam_filter(sms):

    message = re.sub('\W', ' ', sms)
    message = message.lower()
    message = message.split()

    p_spam_given_message = p_spam
    p_ham_given_message = p_non_spam 
    for word in message:
        if word in dict_word_in_spam:
            p_spam_given_message *= dict_word_in_spam[word]
        if word in dict_word_in_non_spam:
            p_ham_given_message *= dict_word_in_non_spam[word]
  
    if p_ham_given_message > p_spam_given_message:
        return 'ham' 
    elif p_ham_given_message < p_spam_given_message:
        return 'spam'
    else:
        return 'check it youself!'

In [155]:
#Let us put the result of Spam Filter analysis into separate column:

In [156]:
testing['Tested by Filter'] = testing['SMS'].apply(spam_filter)
testing.head(10)

Label                                                SMS Tested by Filter
0  spam  Free entry in 2 a wkly comp to win FA Cup fina...             spam
1   ham                I HAVE A DATE ON SUNDAY WITH WILL!!              ham
2  spam  XXXMobileMovieClub: To use your credit, click ...             spam
3   ham  Fine if that’s the way u feel. That’s the way ...              ham
4   ham          Is that seriously how you spell his name?              ham
5   ham  Ahhh. Work. I vaguely remember that! What does...              ham
6   ham  Yeah he got in at 2 and was v apologetic. n ha...              ham
7  spam  07732584351 - Rodger Burns - MSG = We tried to...             spam
8   ham                                WHO ARE YOU SEEING?              ham
9   ham      Didn't you get hep b immunisation in nigeria.              ham

Let us check how the filter works using the secong Testing set:

In [157]:
total = len(testing)
correct = 0
total = len(testing)
    
for row in testing.iterrows():
    row = row[1]
    if row['Label'] == row['Tested by Filter']:
        correct += 1
        
print('Correct:', correct)
print('Incorrect:', total - correct)
print('Accuracy:', round(correct/total*100,2), '%')


Correct: 1090
Incorrect: 24
Accuracy: 97.85 %


# Summary:

A spam filter for SMS messages was built in this project. We used the multinomial Naive Bayes algorithm, and gain 
the accuracy of almost 98 %. 